File: module5_explainability/generate_report.m
function report_data = generate_report(screening_data, output_dir)
% GENERATE_REPORT Primary Module 5 function for XAI map & report generation.
%
% Usage:
%   report_data = generate_report(screening_data)
%   report_data = generate_report(screening_data, 'results/reports')
%
% Inputs:
%   screening_data - Struct containing patient metadata, raw & enhanced images,
%                    IQA results, lesion segmentation masks, and DR grade predictions.
%   output_dir     - (Optional) Directory path to save generated report files.
%
% Output:
%   report_data    - Struct containing generated XAI maps, clinical text,
%                    lesion importance breakdown, summary canvas image, and file paths.

    if nargin < 2 || isempty(output_dir)
        output_dir = fullfile(pwd, 'output_reports');
    end

    if ~exist(output_dir, 'dir')
        mkdir(output_dir);
    end

    % 1. Generate XAI Visual Explanation Maps (Grad-CAM, Saliency, Lesion Composite)
    xai_maps = generate_xai_maps(screening_data);

    % 2. Compute Lesion Importance Breakdown
    dr_grade = screening_data.dr_grading_result.predicted_grade;
    lesion_importance = compute_lesion_importance(screening_data.segmentation_results, dr_grade);

    % 3. Synthesize Natural Language Clinical Text & Recommendations
    clinical_text = generate_clinical_text(screening_data.dr_grading_result, ...
                                           screening_data.iqa_result, ...
                                           lesion_importance);

    % 4. Build 2x3 Diagnostic Summary Canvas Figure
    summary_png_path = fullfile(output_dir, sprintf('Summary_Canvas_%s.png', screening_data.patient_id));
    [fig_handle, summary_canvas_img] = build_summary_figure(screening_data, xai_maps, clinical_text, summary_png_path);
    close(fig_handle);

    % 5. Export Standalone HTML Clinical Report
    html_report_path = fullfile(output_dir, sprintf('Screening_Report_%s.html', screening_data.patient_id));
    export_html_report(screening_data, xai_maps, clinical_text, lesion_importance, html_report_path);

    % 6. Save MAT Structured Summary File
    mat_report_path = fullfile(output_dir, sprintf('Report_Data_%s.mat', screening_data.patient_id));
    
    % Assemble master report_data struct
    report_data = struct(...
        'patient_id', screening_data.patient_id, ...
        'patient_name', screening_data.patient_name, ...
        'eye_side', screening_data.eye_side, ...
        'dr_grade', screening_data.dr_grading_result.grade_label, ...
        'confidence', screening_data.dr_grading_result.confidence, ...
        'iqa_status', screening_data.iqa_result.status, ...
        'xai_maps', xai_maps, ...
        'lesion_importance', lesion_importance, ...
        'clinical_text', clinical_text, ...
        'summary_canvas_img', summary_canvas_img, ...
        'report_files', struct(...
            'html_path', html_report_path, ...
            'summary_png_path', summary_png_path, ...
            'mat_path', mat_report_path ...
        )...
    );

    % Save .mat file
    save(mat_report_path, 'report_data');
end

File: module5_explainability python/module5_explainability python/module5_explainability/src/generate_report.py
import os
import pickle

from generate_xai_maps import generate_xai_maps
from compute_lesion_importance import compute_lesion_importance
from generate_clinical_text import generate_clinical_text
from build_summary_figure import build_summary_figure
from export_html_report import export_html_report

def generate_report(screening_data, output_dir=None):
    """
    Primary Module 5 function for XAI map & report generation.
    """
    if output_dir is None:
        output_dir = os.path.join(os.getcwd(), 'output_reports')
        
    os.makedirs(output_dir, exist_ok=True)
    
    patient_id = screening_data.get('patient_id', 'Unknown')
    
    # 1. Generate XAI Visual Explanation Maps
    xai_maps = generate_xai_maps(screening_data)
    
    # 2. Compute Lesion Importance Breakdown
    dr_grade = screening_data.get('dr_grading_result', {}).get('predicted_grade', 0)
    seg_results = screening_data.get('segmentation_results', {})
    lesion_importance = compute_lesion_importance(seg_results, dr_grade)
    
    # 3. Synthesize Natural Language Clinical Text & Recommendations
    clinical_text = generate_clinical_text(
        screening_data.get('dr_grading_result', {}),
        screening_data.get('iqa_result', {}),
        lesion_importance
    )
    
    # 4. Build 2x3 Diagnostic Summary Canvas Figure
    summary_png_path = os.path.join(output_dir, f'Summary_Canvas_{patient_id}.png')
    fig, summary_canvas_img = build_summary_figure(screening_data, xai_maps, clinical_text, summary_png_path)
    
    # 5. Export Standalone HTML Clinical Report
    html_report_path = os.path.join(output_dir, f'Screening_Report_{patient_id}.html')
    export_html_report(screening_data, xai_maps, clinical_text, lesion_importance, html_report_path)
    
    # 6. Save Structured Summary File
    pkl_report_path = os.path.join(output_dir, f'Report_Data_{patient_id}.pkl')
    
    report_data = {
        'patient_id': patient_id,
        'patient_name': screening_data.get('patient_name', 'N/A'),
        'eye_side': screening_data.get('eye_side', 'N/A'),
        'dr_grade': screening_data.get('dr_grading_result', {}).get('grade_label', 'N/A'),
        'confidence': screening_data.get('dr_grading_result', {}).get('confidence', 0.0),
        'iqa_status': screening_data.get('iqa_result', {}).get('status', 'N/A'),
        'xai_maps': xai_maps,
        'lesion_importance': lesion_importance,
        'clinical_text': clinical_text,
        'summary_canvas_img': summary_canvas_img,
        'report_files': {
            'html_path': html_report_path,
            'summary_png_path': summary_png_path,
            'pkl_path': pkl_report_path
        }
    }
    
    with open(pkl_report_path, 'wb') as f:
        pickle.dump(report_data, f)
        
    return report_data
